Added a trigger delay parameter to delay the CAMERA_FEEDBACK messages (for geotagging purposes)

arjun7965 opened the pull request:
I added a new parameter in the AP_Camera class called "Trigg_Delay" which delays the geotagging packet. Also, added support to feedback_pol such that it also sends a geotagging packet when Feedback_Pol=0. 

I also need your suggestions/advice on how to back-port these changes to Copter-3.5.5 (Stable) as I did not realize master has diverged a lot from Copter-3.5.5.

WickedShell commented on libraries/AP_Camera/AP_Camera.cpp:
```diff
@@ -404,6 +412,11 @@ void AP_Camera::log_picture()
     }
     if (!using_feedback_pin()) {
+        if(_relay_on == _feedback_polarity){
+            hal.scheduler->delay(_trigger_delay);
```
This is absolutely unsafe to do in flight as the entire flight code will be stalled for the duration.

arjun-vinod replied:
Thank you for your quick response. Is there a safer way to delay the Camera feedback message?

WickedShell replied:
You would need to store a timestamp for when you recieved the feedback signal, then check next time you are updated if it has been that long. Can I ask why you want to delay the message though? The event you are trying to tag is happening when the feedback signal is driven, and that is the event one normally is trying to tag.

arjun7965 replied:
Sorry for the late response. I want to delay the CAM_FEEDBACK by few milliseconds so that the picture taken and the geotags are in sync. Thank you for your response

arjun7965 commented:
I rewrote the code so that it uses a counter to delay the CAMERA_FEEDBACK message instead of using 'delay()'.

WickedShell commented:
The build is failing because some of your parameter documentation tags are invalid. (I think it's a capitalization issue, looking at the adjacent ones should solve that).

A couple of implementation comments:

1. Why not just store the timestamp/compare the timestamp when you enter the function again? This gets you out of finicky counter manipulation which we are trying to remove/reduce and will be a problem if we ever poll faster.
1. You originally stated that you wanted to only delay severa milliseconds, however this patch will only be able to delay it in increments of 20 ms as is.
1. Why are you trying to handle rising vs falling edge differently? The falling edge mode is typically used in conjunction with a signal that is high until the camera is triggered, rather then a short pulse where we latch the trailing edge. More importantly `_trigger_duration` is how long we send the command to start taking the photo, rather then how long a pulse we received. It's very common to receive and complete the photo before the full `_trigger_duration` has fired.

arjun7965 commented:
Thank you for your quick response. Regarding your questions: 
1. I implemented it that way because it was consistent with the other code in that module (following the example of _trigger_counter, which was the preexisting counter used in the same blocks where I added _delay_counter).

2. I'm sorry about that confusion. I actually want a delay somewhere around 120 ms and 20ms resolution should work perfectly fine. 

3. This is for when the feedback pin is not being used and when feedback polarity is 0. So, it goes low when the trigger goes high.

arjun7965 commented:
Hi,
Was this discussed in the DevCall? If so, any feedback would be greatly appreciated. Thank you.

OXINARF commented:
@arjun7965 Someone should have left a note here, sorry about that. It's been months ago so I don't really remember the discussion anymore, but here are the notes (https://discuss.ardupilot.org/t/dev-call-may-28-2018-2300-utc/29495/2):

>UTC0011 - https://github.com/ArduPilot/ardupilot/pull/8092
>
>- Do we really want this behaviour?
>- Trying to put a delay in writing the camera packet into the dataflash log
>    - 20ms resolution?
>- Should probably be done post-flight
>- MdB: Camera won’t be that good anyway
>      - Cameras have a huge jitter
>- Better-precision geotagging
>- Trying to fake feedback without having feedback
>- Sometimes it’s worth putting stuff in to make support requests go away
>- Grab later GPS position and attitude information to associate with image
>- Wants 120ms delay
>    - Less that one GPS frame
>    - We could do some extrapolation ourselves in the code
>    - Would give a better result than what’s here
>- Assuming 50Hz loops is bad
>- Millisecond value if it is going to be anything at all
>- CanberraUAV software already interpolates values to give a better position
>    - So does MdB’s software
>- Centiseconds are not a great unit
>    - Already present in code
>- Parameter name isn’t great; it doesn’t delay, it’s more of an offset
>- So… do we want to do this live rather than in post-processing?
>    - Push to get this to a level to be merged?
>      - Can’t go in in current form

arjun7965 commented:
Thank you for your response @OXINARF. I can change this to store timestamps and compare it to offset the feedback message so that we get a much better resolution (like @WickedShell mentioned). or will it be ideal to be done post flight?

davidbuzz commented:
no update since 2018, needs rebase, ppears abandoned

davidbuzz commented:
close pr due to age, and it being discussed at dev call and needs rework.
